aip/0231.md

```diff
@@ -0,0 +1,86 @@
+---
+aip:
+  id: 231
+  state: draft
+  created: 2019-06-18
+permalink: /231
+redirect_from:
+  - /0231
+---
+
+# Batch Get
+
+Some APIs need to allow users to get a specific set of resources at a consistent
+time point (e.g. using a read transaction). A batch get method provides this
+functionality.
+
+## Guidance
+
+Batch get methods are specified using the following pattern:
+
+```proto
+rpc BatchGetBooks(BatchGetBooksRequest) returns (BatchGetBooksResponse) {
+  option (google.api.http) = {
+    get: "/v1/shelves/-/books:batchGet"
+  };
+}
+```
+
+- The RPC's name **must** begin with `BatchGet`. The remainder of the
+  RPC name **should** be the plural form of the resource being retrieved.
+- The request and response messages **must** match the RPC name, with `-Request`
+  and `-Response` suffixes.
+- The HTTP verb **must** be `GET`.
+- The HTTP URI **must** end with `:batchGet`.
+- The URL path **should** represent the collection for the resource, matching
+  the collection used for simple CRUD operations. If the operation spans
+  parents, a dash (`-`) **may** be used as a wildcard.
+- For users with URL length restrictions (e.g. certain old browsers), they can
+  use `POST` to invoke this API method combined with the header
+  `X-HTTP-Method-Override: GET`, which instructs the server to treat the `POST`
+  as a `GET`.
+- There **must not** be a body key in the `google.api.http` annotation.
+
+### Request message
+
+The request for a batch get method **should** be specified with the following
+pattern:
+
+```proto
+message BatchGetBooksRequest {
+  // The requests specifying the books to retrieve.
+  // A maximum of 1000 books can be retreived in a batch.
+  repeated GetBookRequest requests = 1;
```

Comment:
Requiring a `repeated GetBookRequest` forces that level of indirection into the URI string, which seems unfortunate.

    requests.name=foo/1/bar/2&requests.name=foo/1/bar/3

Reply:
Yeah, what does this extra layer of indirection do? No other required fields are allowed, and any other optional fields (e.g. partial response view) should probably apply uniformly across the entire returned set?  

I see it's useful for BatchCreate where client provided IDs would obviously differ per resource... So perhaps the symmetry is ideal...

Reply:
@jgeewax and I discussed this in depth. There are several tradeoffs here.

The big argument for doing it this way is consistency and flexibility -- there _will_ be cases in the future where people really want the ability to augment things on a per-item basis. It will not be the common case, but it will be an advanced case.

The approach Garrett suggests keeps the common case easy (if a bit more verbose than we would like) but it keeps the advanced case possible.

It is worth noting that it does make a couple of anti-patterns become too easy -- e.g. if you had a `dry_run` flag in the individual request, it should not be possible to send five creates with two being dry runs and three being real. Same (probably) for field masks on update, or consistency tokens (anywhere). There are probably some others. So, we probably need some guidance on "hoisting" -- basically, if the general pattern should involve a field always matching within the batch, that field should be hoisted into the batch request and then not be set in the individual requests.

Reply:
I will add a remark about hoisting to each AIP.